File: android/src/main/kotlin/id/oddbit/rectangle_detector/scanner/utility/Utility.java
package id.oddbit.rectangle_detector.scanner.utility;

import android.app.Activity;
import android.content.Context;
import android.content.res.Configuration;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.Matrix;
import android.hardware.Camera;
import android.os.Environment;
import android.util.Log;
import android.view.Surface;
import android.view.View;
import android.widget.Toast;

import java.io.ByteArrayOutputStream;
import java.io.File;

public class Utility {
    private static String TAG = "Utility";

    public static boolean isEmptyOrNull(String value) {
        return (value == null || value.isEmpty());
    }

    public static void showMessage(Context activity, String message) {
        Toast toast = Toast.makeText(activity, message, Toast.LENGTH_SHORT);
        View view = toast.getView();
        toast.show();
    }

    public static int[] getOptimalDimensions(float mediaWidth, float mediaHeight, int width, int height) {
        int layoutWidth = width;
        int layoutHeight = height;
        float ratioWidth = layoutWidth / mediaWidth;
        float ratioHeight = layoutHeight / mediaHeight;
        float aspectRatio = mediaWidth / mediaHeight;
        if (ratioWidth > ratioHeight) {
            layoutWidth = (int) (layoutHeight * aspectRatio);
        } else {
            layoutHeight = (int) (layoutWidth / aspectRatio);
        }
        Log.i(TAG, "layoutWidth: " + layoutWidth);
        Log.i(TAG, "layoutHeight: " + layoutHeight);
        Log.i(TAG, "aspectRatio: " + aspectRatio);
        return new int[]{layoutWidth, layoutHeight};
    }
    public static byte[] rotateImageData(Activity activity, byte[] data, int cameraId) throws Exception {
        Bitmap imageBitmap = null;
        // COnverting ByteArray to Bitmap - >Rotate and Convert back to Data
        if (data != null) {
            imageBitmap = BitmapFactory.decodeByteArray(data, 0, (data != null) ? data.length : 0);
            if (activity.getResources().getConfiguration().orientation == Configuration.ORIENTATION_PORTRAIT) {
                Matrix mtx = new Matrix();
                int cameraEyeValue = setPhotoOrientation(activity, cameraId); // CameraID = 1 : front 0:back
                if (cameraId == 1) { // As Front camera is Mirrored so Fliping the Orientation
                    if (cameraEyeValue == 270) {
                        mtx.postRotate(90);
                    } else if (cameraEyeValue == 90) {
                        mtx.postRotate(270);
                    }
                } else {
                    mtx.postRotate(cameraEyeValue); // cameraEyeValue is default to Display Rotation
                }
                imageBitmap = Bitmap.createBitmap(imageBitmap, 0, 0, imageBitmap.getWidth(), imageBitmap.getHeight(), mtx, true);
            } else {// LANDSCAPE MODE
                //No need to reverse width and height
                Bitmap scaled = Bitmap.createScaledBitmap(imageBitmap, imageBitmap.getWidth(), imageBitmap.getHeight(), true);
                imageBitmap = scaled;
            }
        }
        // Converting the Die photo to Bitmap
        ByteArrayOutputStream stream = new ByteArrayOutputStream();
        imageBitmap.compress(Bitmap.CompressFormat.JPEG, 100, stream);
        return stream.toByteArray();
    }
    private static int setPhotoOrientation(Activity activity, int cameraId) {
        android.hardware.Camera.CameraInfo info = new android.hardware.Camera.CameraInfo();
        android.hardware.Camera.getCameraInfo(cameraId, info);
        int rotation = activity.getWindowManager().getDefaultDisplay().getRotation();
        int degrees = 0;
        switch (rotation) {
            case Surface.ROTATION_0:
                degrees = 0;
                break;
            case Surface.ROTATION_90:
                degrees = 90;
                break;
            case Surface.ROTATION_180:
                degrees = 180;
                break;
            case Surface.ROTATION_270:
                degrees = 270;
                break;
        }

        int result;
        // do something for phones running an SDK before lollipop
        if (info.facing == Camera.CameraInfo.CAMERA_FACING_FRONT) {
            result = (info.orientation + degrees) % 360;
            result = (360 - result) % 360; // compensate the mirror
        } else { // back-facing
            result = (info.orientation - degrees + 360) % 360;
        }

        return result;
    }

    public static File createExternalDirectory(String folder) {
        String var1 = getAppDirectory() + "/" + folder;
        File var2 = new File(var1);
        if (!var2.exists()) {
            var2.mkdir();
        }

        return var2;
    }

    public static String getAppDirectory() {
        String var0 = Environment.getExternalStorageDirectory() + "/CameraPreviewExample";
        File var1 = new File(var0);
        if (!var1.exists()) {
            var1.mkdir();
        }

        return var1.toString();
    }
}

File: android/src/main/kotlin/id/oddbit/rectangle_detector/scanner/utility/Utility.kt
package id.oddbit.rectangle_detector.scanner.utility

import android.app.Activity
import android.hardware.Camera
import android.view.Surface
import java.util.*
import kotlin.math.abs
import kotlin.math.sqrt

object Utility {
    private val TAG = Utility::class.java.simpleName
    private fun compareFloats(left: Double, right: Double): Boolean {
        val epsilon = 0.00000001
        return abs(left - right) < epsilon
    }

    fun determinePictureSize(camera: Camera?, previewSize: Camera.Size): Camera.Size? {
        if (camera == null) return null
        val cameraParams = camera.parameters
        val pictureSizeList = cameraParams.supportedPictureSizes

        pictureSizeList.sortWith(Comparator { size1, size2 ->
            val h1 = sqrt(size1.width * size1.width + size1.height * size1.height.toDouble())
            val h2 = sqrt(size2.width * size2.width + size2.height * size2.height.toDouble())

            h2.compareTo(h1)
        })
        var retSize: Camera.Size? = null

        // if the preview size is not supported as a picture size
        val reqRatio = previewSize.width.toFloat() / previewSize.height
        var curRatio: Float
        var deltaRatio: Float
        var deltaRatioMin = Float.MAX_VALUE
        for (size in pictureSizeList) {
            curRatio = size.width.toFloat() / size.height
            deltaRatio = Math.abs(reqRatio - curRatio)
            if (deltaRatio < deltaRatioMin) {
                deltaRatioMin = deltaRatio
                retSize = size
            }
            if (compareFloats(deltaRatio.toDouble(), 0.0)) {
                break
            }
        }
        return retSize
    }

    fun getOptimalPreviewSize(camera: Camera?, w: Int, h: Int): Camera.Size? {
        if (camera == null) return null
        val targetRatio = h.toDouble() / w
        val cameraParams = camera.parameters
        val previewSizeList = cameraParams.supportedPreviewSizes
        Collections.sort(previewSizeList, Comparator { size1, size2 ->
            val ratio1 = size1.width.toDouble() / size1.height
            val ratio2 = size2.width.toDouble() / size2.height
            val ratioDiff1 = Math.abs(ratio1 - targetRatio)
            val ratioDiff2 = Math.abs(ratio2 - targetRatio)
            if (compareFloats(ratioDiff1, ratioDiff2)) {
                val h1 = Math.sqrt(size1.width * size1.width + size1.height * size1.height.toDouble())
                val h2 = Math.sqrt(size2.width * size2.width + size2.height * size2.height.toDouble())
                return@Comparator h2.compareTo(h1)
            }
            ratioDiff1.compareTo(ratioDiff2)
        })
        return previewSizeList[0]
    }

    fun configureCameraAngle(activity: Activity): Int {
        val angle: Int
        val display = activity.windowManager.defaultDisplay
        angle = when (display.rotation) {
            Surface.ROTATION_0 -> 90 // This is camera orientation
            Surface.ROTATION_90 -> 0
            Surface.ROTATION_180 -> 270
            Surface.ROTATION_270 -> 180
            else -> 90
        }
        return angle
    }
}
